kernel/dev/pci.c

```diff
@@ -0,0 +1,344 @@
+            case 1:
+                *(uint8_t*)target = (uint8_t)value;
+		return;
+            case 2:
+                *(uint16_t*)target = (uint16_t)value;
+		return;
+            case 4:
+                *(uint32_t*)target = (uint32_t)value;
+		return;
+	}
+    );
+}
+
+static void scan_function(uint8_t bus, uint8_t slot, uint8_t func) {
+    struct pci_device* dev = ALLOC(struct pci_device, ALLOC_MISC);
+    *dev = (struct pci_device) {
+        .bus = bus,
+	.slot = slot,
+	.func = func
+    };
+
+    if (pci_readd(dev, 0) == 0xFFFFFFFF) {
+        FREE(dev, ALLOC_MISC);
+	return;
+    }
+
+    uint32_t reg_0 = pci_readd(dev, 0);
+    uint32_t reg_2 = pci_readd(dev, 2 * sizeof(uint32_t));
+
+    dev->device_id = (uint16_t)(reg_0 >> 16);
+    dev->vendor_id = (uint16_t)reg_0;
+    dev->rev_id    = (uint8_t)reg_2;
+    dev->subclass  = (uint8_t)(reg_2 >> 16);
+    dev->pci_class = (uint8_t)(reg_2 >> 24);
+    dev->prog_if   = (uint8_t)(reg_2 >> 8);
+
+    if (dev->pci_class == 0x06 && dev->subclass == 0x04) {
+        // Check if there are more devices hidden behind this bridge
+        uint32_t reg_6 = pci_readd(dev, 6 * sizeof(uint32_t));
+        scan_bus((reg_6 >> 8) & 0xFF);
+    }
+
+    uint16_t sreg = pci_readw(dev, 6);
+    if (sreg & (1 << 4)) {
+        // Traverse the caps list, looking for MSI/MSI-X
+        uint8_t next_off = pci_readb(dev, 0x34);
+
+        while (next_off) {
+            uint8_t ident = pci_readb(dev, next_off);
+
+            switch (ident) {
+                case 5: // MSI compatible
+                    dev->msi_supported = true;
+                    dev->msi_offset = next_off;
+                    break;
+                case 17: // MSI-X compatible
+                    dev->msix_supported = true;
+                    dev->msix_offset = next_off;
+                    break;
+	    }
+
+            next_off = pci_readb(dev, next_off + 1);
+        }
+    }
+
+    VECTOR_PUSH_BACK(&devlist, dev);
+}
+
+static void scan_bus(uint8_t bus) {
+    for (int slot = 0; slot < 32; slot++) {
+        for (int func = 0; func < 8; func++) {
+            scan_function(bus, slot, func);
+        }
+    } 
+}
+
+static void scan_root_bus(void) {
+    struct pci_device root_dev = {0};
+    
+    if (!(pci_readd(&root_dev, 0xC) & 0x800000)) {
+        // Only one PCI host controller
+        scan_bus(0);
+    } else {
+        // Multiple PCI host controllers
+        for (int i = 0; i < 8; i++) {
+            root_dev.func = i;
+
+	    if (pci_readd(&root_dev, 0x0) == 0xffffffff)
+                continue;
+
+	    scan_bus(i);
+	}
+    }
+
+    kernel_print("pci: detected devices:\n");
+    VECTOR_FOR_EACH(&devlist, device,
+        struct pci_device* dev = *device; 
+        kernel_print("  - %02d:%02d:%02d %04x:%04x %02d:%02d:%02d\n", 
+                     dev->bus, dev->slot, dev->func,
+                     dev->vendor_id, dev->device_id,
+                     dev->pci_class, dev->subclass, dev->prog_if);
```

Comment:
can you please not align the arguments with format string here too? just indent them with 4 spaces